### ISSUE TITLE
Structure with "data" is suboptimal for alignment with rails standards

### ISSUE THREAD

**Issue by MartinHinz** [xronos-ch/xronos.rails#165]
We should get rid of the unifying data model, and instead reflect the diversity of our data sources better by using the separate data models directly, joining them in the views or using a abstract class like construction?

**Comment by joeroe** [comment]
I've been experimenting with this. Removing `Data.everything` and instead filtering does make things more idiomatic. It also has the benefit of allowing us to re-use existing filter logic on the individual models, be more precise in the count of return records and speeds things up by eliminating unnecessary joins. There is still a need for the Data model to do some post-processing of filter parameters, but I think this is an okay pattern.

The problem is that, because we use the data browser to essentially filter every model on every other model (e.g. a site that has c14s from a particular c14_lab), we end up joining everything to everything whichever way we do it. `Data.everything` is the way we current do this – but hardcoding the entire data structure into this pseudo-model isn't ideal. Or if we filter each model separately, we have to create associations from each model to every other model (e.g. from site to c14_lab and back) – not ideal either. 

I can't think of an elegant way of doing this, or away around it. Fundamentally, if we want users to be able to "filter everything" from the data browser, there needs to be some sort of model of "everything" in the application.

**Comment by joeroe** [comment]
Looking at how our big brother GBIF handles this gives some ideas for a way out of this problem. They have a search interface (https://www.gbif.org/search) with a text-based multisearch of all data types, but if you click on individual types (e.g. "occurrences", "species") or follow the entries under the "Get data" menu then you get to dedicated search/filter views for each type: https://www.gbif.org/occurrence/search, https://www.gbif.org/species/search.

If we apply this approach to XRONOS' data model, the "data browser" should be dissolved into the index views of each individual model. There we could offer the functionality to filter (or search) by attributes and allow the user to view the results as a table, a map, or a file download. I believe this captures the core logic of the data browser and removes the need to build a complex bespoke view that pulls in data from multiple models.

The downside of course is that if a user wanted to, for example, get C14 *and* dendro dates for a given age range, they would have to perform two searches on two different views. But personally I think this is an acceptable trade-off. As a compromise, we could develop an index view for chrons (since this entity kind of exists conceptually within our data model and could easily be made real) that allows filtering on the common attributes of c14s, typos, dendros, etc. This would be very similar to the current data browser (and could inherit the /data route) but importantly would not include attributes specific to individual chron types (like c14 lab) and therefore should avoid any nasty joins.

**Comment by joeroe** [comment]
> I am not sure about that. Because we rely heavily on the map metaphor. But we can and need to discuss this. For the time being, I made a short term solution for this.
> – @MartinHinz in #388 

I do agree that the map metaphor should be kept central. However, that is still a kind of index view and therefore I think the most Rails-y way to implement it would be to associate it directly with a single model. What I envisage is that we would have multiple map views (e.g. `/dendros?filters=...&view=map` and ``/c14s?filters=...&view=map`) and that these would be displayed to the user as tabs in the data browser, so from the user's perspective they would continue to feel like part of a single interface even though behind the scenes they are separate views.

**Comment by joeroe** [comment]
Some computer-aided brainstorming on a refactored architecture.

---

# Data browser refactor — design document

A composed, multi-model dashboard that shares its filter form, query logic, and table partials with the per-model indexes.

## 1. Headline

A `Filterable` concern lives on the scientific models. Each model declares which of its attributes are filterable, in a small DSL. `Model.filter_by(params[:filter])` returns a scoped relation. The same method is used by:

- the per-model indexes (`sites#index`, `c14s#index`, `typos#index`), which each show only the filters reachable from their model;
- the data browser, which shows the union of all models' filters and composes the four per-model tables in a four-frame Turbo layout.

The data browser is not a special case. Removing it would mean deleting one controller and one layout shell — the model logic, filter form, table partials, and export endpoints are all shared with the per-model indexes.

The `via:` join path on each filter is **inferred from ActiveRecord's reflection graph** by default. The DSL only requires explicit `via:` overrides for unusual cases (self-joins, polymorphic joins where the type can't be auto-derived, performance-critical paths).

## 2. The `Filterable` concern

`app/models/concerns/filterable.rb`. Two responsibilities: a small DSL on the model, and a `Model.filter_by` class method that knows how to apply the relevant filter keys from a flat-but-nested params hash.

### The DSL

```ruby
filter :<field>, type: :<widget_type>, applies_to: :<owning_model>, via: <optional>
```

| Option | Default | Meaning |
|---|---|---|
| `type:` | `:string` | Widget type and SQL operator. `:string`, `:array` (= `IN`), `:range` (= `BETWEEN`). |
| `applies_to:` | `self` | The model that owns the field. Lets filters declared on a model point at fields on related models. |
| `via:` | inferred from AR | Optional. A hash (association chain) or a lambda (custom join). |

In the common case — a field that lives on the declaring model — the declaration is one line:

```ruby
class Site < ApplicationRecord
  include Filterable
  filter :name,         type: :string
  filter :country_code, type: :array
  filter :site_types,   type: :array
  filter :lat,          type: :range
  filter :lng,          type: :range
end
```

For a field that lives elsewhere but is reachable from the declaring model, `applies_to:` redirects the lookup and the join path is computed by reflection:

```ruby
class C14 < ApplicationRecord
  include Filterable
  filter :bp,             type: :range
  filter :lab_identifier, type: :string
  filter :c14_lab,        type: :array
  filter :cal_bp,         type: :range, apply: :with_cal_age_in
  filter :sample_material, applies_to: :sample, key: :material
  filter :site_name,      applies_to: :site,   key: :name
end
```

(The `key:` is the field name on the target model when it differs from the declaration name. If the names match, `key:` is implicit.)

### `Model.filter_by`

```ruby
def self.filter_by(filters, from: model_name.element.to_sym)
  return all if filters.blank?
  filters = filters.deep_stringify_keys
  scope = all
  filters.each do |section, attrs|
    next if attrs.blank?
    attrs.each do |field, value|
      next if value.blank?
      decl = lookup_declaration(section.to_sym, field.to_sym)
      next unless decl
      scope = scope.merge(apply_filter(decl, from, value))
    end
  end
  scope.distinct
end
```

`apply_filter` does one of three things:

1. **Direct attribute** (when `from == decl[:applies_to]`): `where(decl[:column] => value)`.
2. **AR-inferred path**: compute the join path from `from` to the target model via reflections, then `joins(path).where(target => { column => value })`, with any polymorphic type filters auto-injected.
3. **Explicit `via:` lambda**: pass `value` to the lambda and use the returned scope.

### Reflection-path resolution

`Filterable.path(from, to)` performs a BFS over each model's `_reflections`, treating each association as a bidirectional edge. The result is a chain of association names that AR's `joins(...)` can consume directly.

Implementation notes:

- **Both directions of every association are added as edges.** `Site has_many :contexts` and `Context belongs_to :site` are two edges.
- **`has_many :through`** is expanded to its constituent joins.
- **Polymorphic associations** (`has_many :citations, as: :citing`) are detected from the reflection's `options[:as]`. When the path goes through one, `apply_filter` injects `where(citations: { citing_type: <self.name>, citing_id: <self.table_name>.id })`.
- **Memoised at class load time.** The BFS runs once per `(from, to)` pair, not per request.
- **Returns `nil` if no path exists.** `apply_filter` then returns `all` (the filter is silently dropped, which matches the desired behaviour: a root model that can't reach a field ignores that filter).

The BFS can't represent:

- **Self-joins** (a model joining back to itself through a sibling association). These need an explicit `via:` lambda.
- **Conditional joins** (`WHERE EXISTS` rather than `INNER JOIN`). Out of scope for v1; can be added later as another `via:` variant.
- **Performance-sensitive paths** where the shortest path produces worse SQL than a hand-tuned one. The `via:` override is the escape hatch.

### The `apply:` option for custom filters

For filters that don't fit the standard "where column in/between" shape, `apply:` names a scope on the declaring model:

```ruby
class C14 < ApplicationRecord
  filter :cal_bp, type: :range, apply: :with_cal_age_in
  # ...
  scope :with_cal_age_in, ->(tpq_range:, taq_range:) {
    scope = all
    if tpq_range || taq_range
      scope = scope.joins("JOIN cals ON cals.c14_age = c14s.bp AND cals.c14_error = c14s.std")
      scope = scope.where("cals.tpq >= ?", tpq_range.min) if tpq_range
      scope = scope.where("cals.taq <= ?", taq_range.max) if taq_range
    end
    scope
  }
end
```

This is the home for the `cals` join — moved out of the current data browser's bespoke code and into a regular `C14` scope.

### The `via:` lambda override

For paths AR can't express, `via:` accepts a lambda. The lambda receives the filter value and returns a scoped relation:

```ruby
class C14 < ApplicationRecord
  filter :sibling_typo_name,
         applies_to: :typo,
         via: lambda { |name|
           joins(sample: :context)
             .joins("INNER JOIN samples AS sibling_samples ON sibling_samples.context_id = contexts.id")
             .joins("INNER JOIN typos AS sibling_typos ON sibling_typos.sample_id = sibling_samples.id")
             .where(sibling_typos: { name: name })
         }
end
```

This is scenario 1's path: C14 → context → *other* samples in that context → their typos. The reflection graph can't represent it (the "other samples" is the same table as the C14's sample, joined via a self-referential alias), so it's declared explicitly.

## 3. Param shape

Filters are nested one or two levels deep, scoped by the model that owns the field. Rails' `fields_for` produces this naturally:

```ruby
params[:filter] = {
  site:      { name: "Knossos", country_code: ["GR", "TR"], site_types: ["temple"] },
  sample:    { material: ["bone"], taxon: ["sheep"], context: { name: "Layer II" } },
  c14:       { bp: [100, 5000], lab_identifier: "OxA-1234", c14_lab: ["Oxford"], cal_bp: [200, 8000] },
  typo:      { name: "Late Helladic IIIC", approx_start_time: [1200, 1100] },
  reference: { short_ref: "Smith 1999" }
}
```

The tree's depth comes from the model graph (`sample → context → site`), not from an arbitrary root in the data browser. The old `params[:filter][:c14s][:lab_identifier][:name]` shape (model name → field → array of strings, three levels) goes away.

Each model's `filter_by` walks the *entire* hash. For each `(section, field)` pair it knows how to reach, it applies the filter with the appropriate join.

## 4. Per-model declarations (the minimal set)

The complete declaration surface, given the XRONOS data model:

```ruby
# app/models/site.rb
class Site < ApplicationRecord
  include Filterable
  filter :name,         type: :string
  filter :country_code, type: :array
  filter :site_types,   type: :array
  filter :lat,          type: :range
  filter :lng,          type: :range
end

# app/models/sample.rb
class Sample < ApplicationRecord
  include Filterable
  filter :material, type: :array
  filter :taxon,    type: :array
  filter :context,  type: :string
end

# app/models/context.rb — not user-filtered directly; reachable via sample → context
class Context < ApplicationRecord
end

# app/models/c14.rb
class C14 < ApplicationRecord
  include Filterable
  filter :bp,             type: :range
  filter :lab_identifier, type: :string
  filter :c14_lab,        type: :array
  filter :cal_bp,         type: :range, apply: :with_cal_age_in

  scope :with_cal_age_in, ->(tpq_range:, taq_range:) { ... }
end

# app/models/typo.rb
class Typo < ApplicationRecord
  include Filterable
  filter :name,              type: :string
  filter :approx_start_time, type: :range
  filter :approx_end_time,   type: :range
end

# app/models/reference.rb
class Reference < ApplicationRecord
  include Filterable
  filter :short_ref, type: :string
end
```

Five models, ~15 filter declarations, two custom overrides (the `cals` scope and the sibling-typo lambda). The rest of the join logic is inferred.

For completeness, the `sibling_typo_name` declaration (scenario 1) lives on `C14`:

```ruby
class C14 < ApplicationRecord
  filter :sibling_typo_name, applies_to: :typo, via: lambda { |name|
    joins(sample: :context)
      .joins("INNER JOIN samples AS sibling_samples ON sibling_samples.context_id = contexts.id")
      .joins("INNER JOIN typos AS sibling_typos ON sibling_typos.sample_id = sibling_samples.id")
      .where(sibling_typos: { name: name })
  }
end
```

The form's "Typo name" field posts to `params[:filter][:typo][:name]`. `C14.filter_by` walks the `[:typo]` section, finds the `sibling_typo_name` declaration, and applies the lambda. `Site.filter_by` walks the same section and applies its own (different) join to typos. `Typo.filter_by` walks the same section and applies a direct `where(typos: { name: value })`.

## 5. The shared filter form partial

`app/views/shared/_filter_form.html.erb` is rendered by all four controllers. It iterates a `sections:` list of model classes, and for each model iterates the model's filter declarations. Each declaration renders the appropriate input widget.

```erb
<%# app/views/shared/_filter_form.html.erb %>
<%= form_with(url: form_url, method: :get, scope: :filter,
              data: { turbo_frame: turbo_frame }) do |f| %>
  <% sections.each do |model| %>
    <section class="filter-section">
      <h3><%= model.label.pluralize.humanize %></h3>
      <% model.filter_declarations.each do |decl| %>
        <%= render "shared/filter_field", form: f, model: model, decl: decl %>
      <% end %>
    </section>
  <% end %>
  <%= f.submit "Apply filters", class: "btn btn-primary" %>
<% end %>
```

`app/views/shared/_filter_field.html.erb` dispatches on `decl[:type]`:

- `:string` → text input
- `:array` → Tom Select with `multiple: true`, posting a comma-joined string
- `:range` → noUiSlider with hidden min/max inputs

The controllers pass:

| Controller | `sections:` |
|---|---|
| `sites#index` | `[Site, Reference]` |
| `c14s#index` | `[C14, Sample, Site, Reference]` |
| `typos#index` | `[Typo, Sample, Site, Reference]` |
| `data_browser#index` | `[Site, Sample, C14, Typo, Reference]` |

A new filter is added by adding one line to the model. The form, the strong-params, and the query all pick it up automatically.

## 6. The per-model indexes

Each per-model controller becomes a thin pass-through:

```ruby
class C14sController < ApplicationController
  def index
    @c14s = C14.filter_by(filter_params).includes_for_table
    respond_to do |format|
      format.html { pagy_render(@c14s) }
      format.json { ... }
      format.csv  { ... }
    end
  end

  private

  def filter_params
    params.fetch(:filter, {}).permit!
  end
end
```

The filter form (the shared partial) is rendered into the index view, gated on a `?filter` query param being present (default unfiltered view is a plain table, no sidebar).

The per-model `_table.html.erb` partials stay almost as they are. The one change: the `unless controller_name == "sites"` guard becomes `unless local_assigns.fetch(:show_site, true) == false` — explicit, no implicit controller-name check, defaults to the current behaviour.

## 7. The data browser

The data browser is a layout shell. One controller, four turbo-frames, four panel endpoints, one GeoJSON endpoint.

### Routes

```ruby
# config/routes.rb
get "/data"                          => "data_browser#index"
get "/data/sites"                    => "data_browser#sites"
get "/data/c14s"                     => "data_browser#c14s"
get "/data/typos"                    => "data_browser#typos"
get "/data/sites_map"                => "data_browser#sites_map"
get "/data/sites_map.geojson"        => "data_browser#sites_map_geojson"
```

### Layout

```erb
<%# app/views/data_browser/index.html.erb %>
<%= render "shared/filter_form", sections: [Site, Sample, C14, Typo, Reference],
                                 form_url: data_path,
                                 turbo_frame: active_panel_frame %>

<%= turbo_frame_tag "sites_map_frame", src: sites_map_path(filter: params[:filter]) do %>
  <p>Loading map…</p>
<% end %>

<nav class="nav nav-pills">
  <%= link_to "Sites",   sites_data_path(filter: params[:filter]),  data: { turbo_frame: "sites_panel" } %>
  <%= link_to "C14s",    c14s_data_path(filter: params[:filter]),   data: { turbo_frame: "c14s_panel" } %>
  <%= link_to "Typos",   typos_data_path(filter: params[:filter]),  data: { turbo_frame: "typos_panel" } %>
</nav>

<%= turbo_frame_tag "sites_panel", src: sites_data_path(filter: params[:filter]) do %>
  <p>Loading sites…</p>
<% end %>
<%= turbo_frame_tag "c14s_panel", src: c14s_data_path(filter: params[:filter]) do %>
  <p>Loading C14s…</p>
<% end %>
<%= turbo_frame_tag "typos_panel", src: typos_data_path(filter: params[:filter]) do %>
  <p>Loading typos…</p>
<% end %>
```

### Controller

```ruby
class DataBrowserController < ApplicationController
  def index
    # renders the layout shell
  end

  def sites
    @sites = Site.filter_by(filter_params).includes_for_table
    @pagy, @sites = pagy(:countish, @sites, page_param: :sites_page)
    render partial: "data_browser/panel_sites"
  end

  def c14s
    @c14s = C14.filter_by(filter_params).includes_for_table
    @pagy, @c14s = pagy(:countish, @c14s, page_param: :c14s_page)
    render partial: "data_browser/panel_c14s"
  end

  def typos
    @typos = Typo.filter_by(filter_params).includes_for_table
    @pagy, @typos = pagy(:countish, @typos, page_param: :typos_page)
    render partial: "data_browser/panel_typos"
  end

  def sites_map
    # renders the map partial with markers_url local
  end

  def sites_map_geojson
    @sites = Site.filter_by(filter_params).where.not(lat: nil, lng: nil)
    render geojson: @sites
  end

  private

  def filter_params
    params.fetch(:filter, {}).permit!
  end
end
```

Each panel partial is essentially:

```erb
<%# app/views/data_browser/panel_c14s.html.erb %>
<%= turbo_frame_tag "c14s_panel" do %>
  <%= render "c14s/table", c14s: @c14s, show_site: true %>
  <%== @pagy.series_nav(:bootstrap) %>
<% end %>
```

The **same** per-model `_table.html.erb` that `c14s#index` already uses, fed a filtered scope. The download dropdown (CSV / JSON / MIAARD JSON) is rendered alongside, with links that pass `params[:filter]` through.

### Frame refresh

When the user clicks "Apply filters":

- The form posts to `data_path` with `data-turbo-frame="<active panel>"`.
- Only the active panel frame reloads. The other three keep their cached state.
- The map frame also reloads, because its `src:` is parameterised on `params[:filter]` and the browser URL has changed.

When the user clicks a different tab:

- The new tab's `src:` already includes the current `params[:filter]`, so it loads with up-to-date data.
- The previously-active tab's content stays cached.

## 8. Exports

Each per-model controller handles its own exports. The data browser's download dropdown just links to these URLs with the current `params[:filter]` passed through.

| Format | Endpoint | Source |
|---|---|---|
| Sites CSV | `GET /sites.csv?filter=…` | `sites#index` `format.csv` |
| Sites GeoJSON | `GET /data/sites_map.geojson?filter=…` | `data_browser#sites_map_geojson` |
| Sites JSON | `GET /sites.json?filter=…` | `sites#index` `format.json` |
| C14s CSV | `GET /c14s.csv?filter=…` | `c14s#index` `format.csv` |
| C14s JSON | `GET /c14s.json?filter=…` | `c14s#index` `format.json` |
| C14s MIaaRD JSON | `GET /c14s.json?schema=miaard&filter=…` | `c14s#index` `format.json` |
| Typos CSV | `GET /typos.csv?filter=…` | `typos#index` `format.csv` |
| Typos JSON | `GET /typos.json?filter=…` | `typos#index` `format.json` |

The data browser doesn't reimplement any export logic. Each per-model controller already has a `format.csv` and `format.json` block; they just call `Model.filter_by(params[:filter])` instead of their previous hand-written `where(model_params)`.

## 9. What gets deleted

- `app/models/xronos_data.rb` (208 LOC) — superseded by `Filterable#filter_by` on each model.
- `app/controllers/xronos_data_controller.rb` (275 LOC) — superseded by `DataBrowserController` and the per-model controllers.
- `app/views/xronos_data/` directory entirely — `_filter_fields.html.erb` is replaced by the shared filter form; `_map.html.erb` is replaced by `data_browser/_sites_map.html.erb`; `_table.html.erb` is replaced by the per-model `c14s/_table.html.erb`; the toolbar and filter-status are absorbed into the data browser layout.
- `config/routes.rb:95,96,103` — the dangling `data/index` GET/POST and `/resetfilter`.
- `config/schedule.rb:30,36` — the broken `refreshers:data_views` and `Data.store_data_as_json` entries.
- The custom v1 API write actions in `app/controllers/api/v1/xronos_data_controller.rb` (read endpoints can stay until v1 deprecation).
- The v1 API's unused jbuilder partials under `app/views/api/v1/`.
- The unfiltered cache (the keys under `xronos_data/unfiltered/v2/*`); `:countish` already memoises per-page counts.

## 10. Worked examples

### Scenario 1: C14s from contexts that also have a typo "Neolithic"

User fills in the "Typo name" field under the Typo section: `params[:filter] = { typo: { name: "Neolithic" } }`.

The C14 panel's `filter_by` walks `params[:filter][:typo]`, finds the `sibling_typo_name` declaration on `C14`, and applies the lambda:

```sql
SELECT DISTINCT c14s.*
FROM c14s
INNER JOIN samples AS c14_sample ON c14_sample.id = c14s.sample_id
INNER JOIN contexts ON contexts.id = c14_sample.context_id
INNER JOIN samples AS sibling_samples ON sibling_samples.context_id = contexts.id
INNER JOIN typos AS sibling_typos ON sibling_typos.sample_id = sibling_samples.id
WHERE sibling_typos.name = 'Neolithic'
```

The Sites panel's `filter_by` (also walking `[:typo]`) applies its own (different) join:

```sql
SELECT DISTINCT sites.*
FROM sites
INNER JOIN contexts ON contexts.site_id = sites.id
INNER JOIN samples ON samples.context_id = contexts.id
INNER JOIN typos ON typos.sample_id = samples.id
WHERE typos.name = 'Neolithic'
```

The Typo panel's `filter_by` (walking the same section) applies a direct filter:

```sql
SELECT DISTINCT typos.*
FROM typos
WHERE typos.name = 'Neolithic'
```

Same form field, three different joins, each panel produces the right result for its root model.

### Scenario 2: Sites with C14s having taxon *Homo sapiens*

User fills in "Taxon" under the Sample section: `params[:filter] = { sample: { taxon: "Homo sapiens" } }`.

The Sites panel's `filter_by` walks `params[:filter][:sample]`, looks up the `taxon` declaration on `Sample`, computes the AR-inferred path `Site → contexts → samples → taxons`, and applies it:

```sql
SELECT DISTINCT sites.*
FROM sites
INNER JOIN contexts ON contexts.site_id = sites.id
INNER JOIN samples ON samples.context_id = contexts.id
INNER JOIN taxons ON taxons.id = samples.taxon_id
WHERE taxons.name = 'Homo sapiens'
```

The C14 panel's `filter_by` (walking the same section) computes the path `C14 → sample → taxon` and applies it:

```sql
SELECT DISTINCT c14s.*
FROM c14s
INNER JOIN samples ON samples.id = c14s.sample_id
INNER JOIN taxons ON taxons.id = samples.taxon_id
WHERE taxons.name = 'Homo sapiens'
```

The Typo panel does the same with `Typo → sample → taxon`.

### Scenario 3: Typos cited to reference Smith 1999

User fills in "Short ref" under the Reference section: `params[:filter] = { reference: { short_ref: "Smith 1999" } }`.

The Typo panel's `filter_by` walks `params[:filter][:reference]`, looks up the `short_ref` declaration on `Reference`, computes the AR-inferred path `Typo → citations → references` (noting that the `citations` association is polymorphic), and applies it with the auto-injected type filter:

```sql
SELECT DISTINCT typos.*
FROM typos
INNER JOIN citations ON citations.citing_id = typos.id AND citations.citing_type = 'Typo'
INNER JOIN references ON references.id = citations.reference_id
WHERE references.short_ref = 'Smith 1999'
```

The Sites and C14 panels produce analogous queries with `citing_type = 'Site'` and `citing_type = 'C14'` respectively — all inferred, none declared.

## 11. Tradeoffs and risks

| Risk | Mitigation |
|---|---|
| Silent behaviour change when an association is added/removed. | Unit test per filter declaration asserting the SQL it produces. `Filterable.paths` introspection method for debugging. |
| BFS picks the "wrong" path when multiple paths exist. | `via:` override available per filter. The BFS is deterministic; surprising choices are stable and can be overridden. |
| `has_many :through` paths can produce expensive SQL. | The BFS picks shortest-path, but `via:` lets the model author pick a more efficient path when needed. |
| Self-joins (scenario 1) need explicit declaration. | Acceptable — they are an edge case. One lambda per such filter. |
| The filter form on per-model indexes gets bigger (4-5 sections each). | The per-model indexes can opt out of the Reference section if it's too noisy. The data browser always shows all five. |
| The polymorphic `citing_type` filter is auto-injected silently. | Unit test for each filter that hits a polymorphic path. |
| Memoising paths at class load time means changing an association requires a server restart. | Acceptable — associations are not changed at runtime. The path cache is rebuilt on process restart. |

## 12. Open implementation details

To be settled at implementation time, not design time:

- **Exact BFS implementation.** Whether to use `ActiveRecord::Reflection` directly or build a parallel graph. Reflection-based is simpler; parallel-graph is more explicit and testable.
- **Path cache representation.** Hash of `{ from => { to => path } }` per model, computed at class load.
- **Strong params strategy.** `permit!` for simplicity (the form's `fields_for` only produces known keys), or a per-model permit list derived from the filter declarations.
- **The `includes_for_table` scope on each model.** A small addition to capture the eager-load the existing `_table` partials need.
- **The `show_site:` local default.** `true` for backwards compat with `sites#show`'s embedding; `false` when called from the data browser (since the Sites column is redundant when the data browser is showing all four panels).
- **Filter form visibility on per-model indexes.** Whether the sidebar is always shown (collapsed by default if no filter is active) or only when a filter is present.

## 13. Phased delivery

Roughly in order; each phase is reviewable on its own:

1. **Foundations.** `Filterable` concern, AR reflection-path resolver, per-model filter declarations, `C14.with_cal_age_in` scope, unit tests for each declaration and each scenario.
2. **Per-model indexes.** Each controller calls `Model.filter_by(params[:filter])`. The shared filter form partial renders the per-model sections. No data browser change yet.
3. **Data browser.** New `DataBrowserController`, four-frame layout, panel endpoints, generic map partial with `markers_url:` local.
4. **Exports.** GeoJSON moves to `data_browser#sites_map_geojson`. The other exports continue to live on the per-model controllers; the data browser's download dropdown links to them.
5. **Deletion.** `xronos_data*` files, dangling routes, broken cron, v1 API write actions, unfiltered cache.
6. **Polish.** Per-index filter form visibility, `CHANGELOG.md` entry, public docs / README update.

## 14. Alternative: RESTful `DataBrowserController`

The design in sections 7 and 8 has `DataBrowserController` with non-standard actions (`sites`, `c14s`, `typos`, `sites_map`, `sites_map_geojson`). A stricter interpretation of REST rejects those — they're not standard resource actions. The data browser is a *view* that composes multiple resources, not a resource itself, so a fully RESTful shape is achievable by reducing `DataBrowserController` to a single `index` action and serving the panels from the existing per-model controllers.

### The principle

- **`DataBrowserController` has one action: `index`.** Renders the shell.
- **Each panel is a turbo-frame whose `src:` is a per-model URL.** `sites_path`, `c14s_path`, `typos_path`.
- **The per-model controllers' `index` actions serve both** the full page (when accessed directly) and the panel (when loaded as a turbo-frame). The branching is on `turbo_frame_request?` plus the requested frame ID.

The non-standard panel actions on `DataBrowserController` all go away. The map's GeoJSON is just `sites_path(format: :geojson)`, served by the existing `sites#index` (which gains a new `format.geojson` block).

### Routes

```ruby
# config/routes.rb
resources :sites
resources :c14s
resources :typos

get "/data" => "data_browser#index"
```

The entire data browser routing. Everything else comes from the existing resources.

### `DataBrowserController`

```ruby
# app/controllers/data_browser_controller.rb
class DataBrowserController < ApplicationController
  def index
    # Renders the shell.
  end
end
```

A 6-line controller. No non-standard actions, no `c14s`/`typos`/`sites` methods. The seven standard RESTful actions don't apply (no record, no create/update/destroy), so the controller has just `index`. The data browser is a *view*, not a resource.

### The shell

```erb
<%# app/views/data_browser/index.html.erb %>
<%= render "shared/filter_form", sections: [Site, Sample, C14, Typo, Reference],
                                 form_url: data_path, turbo_frame: "active_panel" %>

<%= turbo_frame_tag "sites_map_frame", src: sites_path(view: :map, filter: params[:filter]) do %>
  <p>Loading map…</p>
<% end %>

<nav class="nav nav-pills">
  <%= link_to "Sites",   sites_path(filter: params[:filter]),  data: { turbo_frame: "sites_panel" } %>
  <%= link_to "C14s",    c14s_path(filter: params[:filter]),   data: { turbo_frame: "c14s_panel" } %>
  <%= link_to "Typos",   typos_path(filter: params[:filter]),  data: { turbo_frame: "typos_panel" } %>
</nav>

<%= turbo_frame_tag "sites_panel", src: sites_path(filter: params[:filter]) do %>
  <p>Loading sites…</p>
<% end %>
<%= turbo_frame_tag "c14s_panel", src: c14s_path(filter: params[:filter]) do %>
  <p>Loading C14s…</p>
<% end %>
<%= turbo_frame_tag "typos_panel", src: typos_path(filter: params[:filter]) do %>
  <p>Loading typos…</p>
<% end %>
```

The frames' `src:`s are the canonical per-model URLs. Turbo loads them concurrently.

### The per-model controller handles both modes

The `index` action in `SitesController` (and `C14sController`, `TyposController`) detects the turbo-frame context and renders the appropriate partial:

```ruby
class SitesController < ApplicationController
  def index
    @sites = Site.filter_by(filter_params).includes_for_table
    @pagy, @sites = pagy(:countish, @sites)

    respond_to do |format|
      format.html do
        if turbo_frame_request?
          render panel_partial_for(turbo_frame_request_id),
                 locals: panel_locals
        else
          render :index
        end
      end
      format.geojson { render geojson: @sites }
      format.json
      format.csv
    end
  end

  private

  def panel_partial_for(frame_id)
    case frame_id
    when "sites_map_frame"  then "sites/map_panel"
    when "sites_panel"      then "sites/table_panel"
    end
  end

  def panel_locals
    { sites: @sites, pagy: @pagy,
      markers_url: sites_path(format: :geojson, filter: filter_params) }
  end
end
```

`turbo_frame_request?` and `turbo_frame_request_id` are both provided by `turbo-rails`. The branching is small and explicit. The same pattern repeats in `C14sController` and `TyposController` — one `case` statement per controller, mapping frame IDs to partials.

### The panel partials

```erb
<%# app/views/sites/_table_panel.html.erb %>
<%= render "sites/table", sites: sites, pagy: pagy, show_site: true %>
<%== pagy.series_nav(:bootstrap) %>

<%# app/views/sites/_map_panel.html.erb %>
<div data-controller="map"
     data-map-target="container"
     data-map-markers-url-value="<%= markers_url %>">
</div>
```

These are the same partials the data browser shell renders, just extracted. The `c14s/_table_panel.html.erb` and `typos/_table_panel.html.erb` follow the same shape.

### The filter form

The form in the shell targets whatever frame is active. Active-frame state lives in a small Stimulus controller:

```javascript
// app/javascript/controllers/data_browser_tabs_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["form", "frame"]

  connect() {
    this.updateTarget()
  }

  activate(event) {
    this.frameTargets.forEach(f => f.dataset.active = "false")
    event.currentTarget.dataset.active = "true"
    this.formTarget.action = event.currentTarget.dataset.formAction
    this.formTarget.dataset.turboFrame = event.currentTarget.dataset.turboFrame
  }

  updateTarget() {
    const active = this.frameTargets.find(f => f.dataset.active === "true") || this.frameTargets[0]
    if (active) {
      this.formTarget.action = active.dataset.formAction
      this.formTarget.dataset.turboFrame = active.dataset.turboFrame
    }
  }
}
```

When the user clicks a tab, the form's `action` and `data-turbo-frame` are updated to match. Submitting the form then updates the active frame; the others keep their cached state. The form's `action` per tab is the per-model URL of that tab's panel.

### Request flow in practice

- **`GET /data`** → `data_browser#index` → the shell with empty turbo-frames.
- **The browser fetches** `sites_path(view: :map)`, `sites_path`, `c14s_path`, `typos_path` in parallel, each with `Turbo-Frame: <id>` headers.
- **Each per-model controller** detects the turbo-frame, queries the data, and returns the appropriate panel partial.
- **The user clicks Apply.** The form posts to the active tab's per-model URL with `data-turbo-frame` set. The active frame reloads with the new filter applied; the others keep their cached state.
- **The user clicks a different tab.** Turbo loads that tab's per-model URL into its frame, with the current filter from the URL bar already applied.
- **GeoJSON** is served by `sites#index`'s `format.geojson` block at `GET /sites.geojson?filter=…`. The exports table in section 8 is updated accordingly: the GeoJSON row's source becomes `sites#index` `format.geojson`.

### Tradeoffs

**Pros:**

- `DataBrowserController` is a 6-line file with one action. Truly RESTful.
- The per-model controllers are the single source of truth for their data; the data browser just composes their views.
- Adding a new panel (e.g. References) is a single new `references_path` in the shell and a new branch in `ReferencesController#index`.
- The map and the table share the same controller action and the same `format.geojson` — there's no duplicated query.

**Cons:**

- Each per-model controller now has a `case` statement mapping frame IDs to partials. The data browser and the per-model controllers are coupled by the frame ID convention.
- The filter form's `action` is dynamic (set by Stimulus), which is more JS than a static form.
- A non-data-browser turbo-frame request to `sites#index` (if any other view uses it) would hit the same panel-rendering code. The `case` statement is the safeguard; if the frame ID isn't recognised, fall back to the full index.

### Final shape of the per-model controller

Same for each of the three:

```ruby
class SitesController < ApplicationController
  def index
    @sites = Site.filter_by(filter_params).includes_for_table
    @pagy, @sites = pagy(:countish, @sites)

    respond_to do |format|
      format.html do
        if turbo_frame_request?
          render panel_partial_for(turbo_frame_request_id),
                 locals: { sites: @sites, pagy: @pagy,
                           markers_url: sites_path(format: :geojson, filter: filter_params) }
        else
          render :index
        end
      end
      format.geojson { render geojson: @sites }
      format.json
      format.csv
    end
  end
end
```

Three controllers, three near-identical methods, three `case` statements. The repetition can be lifted into a `Panelable` concern if it bothers you; or left as-is given how small it is.
